### ISSUE TITLE
[Discussion] Some thoughts I've had on the upcoming Air Canister overhaul

### ISSUE THREAD

**Issue by HipHopHuman** [TeamPneumatic/pnc-repressurized#890]
I asked about this the other day [in the Discord](https://discord.com/channels/380076480241336323/380100722420613121/878556059441176607) because I had some concerns that, while a cool idea on paper, seems like it would not offer that much value once players reach the Aerial Interface stage.

You responded with pretty much the same concerns, so I've been brainstorming on possible ideas to make that more convenient and this is my thinking. Note that the following bullet points don't necessarily have to be grouped together as combined functionality, they are just ideas that I'm listing in isolation from each other.

- Make Air Canisters behave more like fluid tanks, in that they can be combined in the crafting grid to get bigger capacities. There will still be two tiers of canister, but one can "upgrade" each tier, so you could have a 10.0 bar canister that can store 15kml of air, a 10.0 bar that can store 30kml of air and a 10.0 bar that can store 60kml of air, and the same for the 20.0 bar tier. This would work the same way the other Pneumatic tools do, where they can have an air volume that far exceeds their base air storage.
- Introduce a new craftable item, the Air Canister Array. It can hold up to 4 Air Canisters in a single inventory slot.
- Allow Air Canisters to accept Volume Upgrades in the Charging Station.
- Allow Air Canisters to be enchanted with Unbreaking, similar to how Minigun Ammo can be enchanted with Unbreaking. This will give it a chance to not consume air per tick. Up to 30% chance at Unbreaking III.
- Allow Air Canisters to be enchanted with Holding (if Thermal is installed) to increase their base air capacity.
- Instead of offering just *one* active mode where the Air Canister supplies air to all items in the inventory, offer multiple modes (cycle through them with shift+right click). Giving the player more choice as to which items get supplied with air will help them make their Air Canisters last a little longer:
  - Regular Mode: The Air Canister will supply all other items, equipped or in the inventory with air. It will not take air from any items.
  - Ignore Inventory Mode: The Air Canister will supply air to only the worn armor pieces and items in the hotbar. It will not take air from any items.
  - Prefer Inventory Mode: The Air Canister will supply air to only the items in the main inventory. It will not take air from any items.
  - Prefer Hotbar Mode: The Air Canister will supply air to only the items in the hotbar. It will not take air from any items.
  - Prefer Armor Mode: The Air Canister will supply air to only the armor/equipment worn by the player. It will not take air from any items.
  - Prefer Held Item Mode: The Air Canister will supply air to *only* the actively held item. It will not take air from any items.
  - Equalize Mode: The Air Canister will supply air to all items, but it *will* take air from items in order to ensure that air pressure is equalized across all items. This mode will essentially turn the Air Canister into a catalyst instead of a battery, where all of the air pressure bars on items combine into one total air pressure bar that is shared among all items with their air capacities combined. This should make the basic tier Air Canister offer far more value to the player, as *every* pressurized item in their inventory essentially acts as one big battery. For example, if I had an Air Canister in this mode, as well as a Pneumatic Wrench and a Minigun, all of which are at 5.0 bar, and I shoot the Minigun until it ticks down to 4.7 bar, then the Air Canister will take air from itself and from the Pneumatic Wrench to charge the Minigun, such that the Pneumatic Wrench, the Minigun and the Air Canister all sit at 4.9 bar. The challenge in implementation here is figuring out what to do when the player has two separate tiers of items in their inventory at the same time. Perhaps a basic tier Air Canister should always ignore advanced tier items if the basic tier items are at maximum capacity?

This is of course all in *addition* to the introduction of a tier 1.5 compressor, as you have already discussed on the roadmap, as being able to reach 10.0 bar will certainly make a difference pre-advanced stage.

What are your thoughts?

**Comment by desht** [comment]
Generally, yes - I like most of these ideas.

* It's quite a lot of modes to cycle through.  I was thinking more of "off", "hotbar", "armor", or "all" (and I guess "all+balance")
* I don't like the Unbreaking idea, sorry.  Feels like getting air from nowhere (and Unbreaking is for durability, not energy...)

Rather than adding all the functionality to Air Canisters themselves, though, it might be better to require crafting of the array you mentioned, and make that the item you use for charging your inventory - 4 air canisters give a base volume of 12,000mL (so 60,000mL at 5 bar or 120,000mL at 10 bar), which is decent storage early on.  Add in Volume Upgrades and you have a lot of air (even more if Holding is available).  That way, Air Canisters remain mostly just a crafting component.

My concern remains that this all becomes basically obsolete once you get the chestplate & charging upgrades.  Though I suppose in an expert pack, that could be a long way past getting the Charging Array, or whatever it's called.

Also gonna need Rid to do a nice model for this array :)

**Comment by HipHopHuman** [comment]
Yeah, that's why I mentioned that these are all isolated ideas. I quite liked the array idea too, although I hadn't considered that this item could be ugraded itself - which just makes a lot of sense. The only thing I didn't like about the charging array idea was the name, but I couldn't settle on a better one. Though, a quick little google search led me to this page: https://metalsub.com/equalizers

Which makes me think of "Equalizer" - it makes sense, logically speaking, at least to me, and sounds a lot better than "Air Canister Array" or "Charging Array". 

As for your concerns about it being made obsolete once one gets the chestplate and charging upgrades - honestly, I have *some* of that concern as well, but to be honest, I think that that doesn't matter so much. In a bare PNC playthrough (i.e. without Thermal and it's Holding enchant), I certainly ran into low-pressure territory when doing mining expeditions with the jackhammer, even with the chestplate and a lot of Volume Upgrades. 

Sometimes I didn't even have the resources to fully max out the charging upgrades and I definitely did have to return to the surface when charge was getting a bit too low for comfort, but not always.

I would see this as a more situational thing. Like, it's there for you to use *if* you've been a little unlucky gathering Lapis in the early game, or you're playing an expert mode pack, but if you are lucky and you don't need it, then more power to you...?
